### ISSUE TITLE
[question] Does Conan look for the latest RREV for the latest RREV for the current profile?

### ISSUE THREAD

**Issue by timbrown5** [conan-io/conan#7060]
Background:
We are trying to use Conan to support multiple OSs. Previously, we found there we had conflicts when trying to host both CentOS and Ubuntu builds of common dependencies (like boost), so we added a profile which specifies `conan.settings.os.distro=centos_x_x` and `conan.settings.os.distro=ubuntu_x_x`.
This has enabled us to have multiple builds of the same RREV (we use SCM revision mode) that target different OS/Distros.

Now we are seeing issues one package has been built, say for Ubuntu, but the build for another OS, say CentOS is stll running. In this example:
 - If we do `conan install` on Ubuntu machine, it works as expected.
 - If we do `conan install` on CentOS machine, it fails when trying to get a revision.

It seems to be looking for the RREV for the newest build (Ubuntu), but as there is no CentOS build for that RREV yet it (rightly) falls.

My question is:
 - Is the above the expected behaviour - that Conan will try and get all builds from the latest RREV, even if there is no build for this RREV for the current OS/Profile?

- [ ] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Yes, this is expected behavior. The thing is that it is not possible or even desirable to block or atomic lock a recipe in a server that can add more binaries later. Revisions will guarantee that you are getting the thing that you want, if there were changes in the SCM and it was uploaded with a new RREV, then yes, trying to install package binaries for it that haven't been uploaded yet, will fail. Using old packages that are outdated from the last source code changes, is undesired.


This is a relatively common "devops" problem, that is solved in this way:

- Different remotes are used. For example it is easy to set up 2 different remotes in Artifactory, lets call them "develop" and "build".
- The CI that builds packages for new commits get dependencies from "develop" repo to build things. But new packages are uploaded to the "build" repo
- Developer at their machine use the "develop" repo to install things. Only if they want to test or verify some packages of some platform that are being built, they can opt-in to get artifacts from the "build" repo (typically a exceptional or debug-only use case)
- Once the build of the new changes (new SCM commit, new RREV) is completely finished, for all distros/platforms/configurations, it can be "promoted" to the "develop" repo.
- There are mechanisms to do this, it is very recommended to use Artifactory mechanisms, like the Rest API (can do copies, or even complete BuildInfo promotions, of complete dependency graphs (a bit wip for Conan yet). This makes it very fast, as Artifactory implements artifact de-duplications, so it becomes a metadata copy only.
- There are ways you can control and annotate the artifacts and the different platforms that are created. A possible approach is to store the lockfiles of every platform build in a generic repo, in a folder that is named after the CI unique job number. If the promotion mechanism is decoupled from the actual build of the different platforms, you can verify that the full thing is ready if you find all the expected lockfiles in this folder.

Please let us know if this helps.

**Comment by timbrown5** [comment]
Thank you for the quick response.
As you say, we can work around our problem adding an overarching pipeline which does:
```
boost -+--> Ubuntu  --+-> promote
       |--> Windows --|
       |--> CentOS  --|
```
We were thinking of something similar, but with the last step as upload. From your description, promote seems the better option.
This approach does seem to solve another headache we were conscious of:
 - if the Ubuntu build picked up a dependency at RREV1 and CentOS at RREV2.
 - If REV1 and REV2 had significant differences, say incompatible ABIs, one of the builds could fail and the other pass. In this case, it wouldn't be obvious that the builds were using different versions of the dependency.

**Comment by memsharded** [comment]
> We were thinking of something similar, but with the last step as upload. From your description, promote seems the better option.

Yes, upload can take a while, and in the meantime you will temporarily have incomplete packages. Doing it via Artifactory will be way faster.

> if the Ubuntu build picked up a dependency at RREV1 and CentOS at RREV2.

Yes, exactly, the idea is that when you launch a build, you build the lockfiles from the "develop" repo, for all your configurations. Then you launch them, using those lockfiles to ensure that later possible changes to "develop" during the build doesn't change the build. The likelihood of inconsistencies will be drastically reduced. And for full guarantee, you could even mutually lock the process of capturing lockfiles and the process of running promotions in CI.

**Comment by timbrown5** [comment]
I spoke to some of our devs and they would like the option of testing a build with a new Ubuntu package version, even is another us (say Mac/Windows) fails.

I have looked to work around this by using `conan search xxx -q yyy` where `yyy` is the details of my profile to get the REV for the current system.
The problem I can up against is that `conan install` doesn't support a package revision with a RREV.

The only way I can think of working around this is generating the `conan.lock` file myself, which seems a little extreme.

**Comment by timbrown5** [comment]
> I have looked to work around this by using `conan search xxx -q yyy` where `yyy` is the details of my profile to get the REV for the current system.
> The problem I can up against is that `conan install` doesn't support a package revision with a RREV.
>
> The only way I can think of working around this is generating the conan.lock file myself, which seems a little extreme.

Someone in my team pointed out that the above comment is wrong and it turns out I was using the wrong reference. I was using `Package ID:` (returned by `conan search)` instead of RREV (which is returned by `conan search -rev`) - I still often get them confused.

So to install the latest package which matches a given profile, you can do:
```
conan profile show <name>
# This will list the details of the profile, which are needed as input into the `-q` parameter in the search below.
conan search <conan_reference> -r conan_entry.repo_name -rev -q "<details of profile>"
# this will return you the available RREVs for your profile.
conan install <conan_reference>#<RREV>
```
As a side note, it would be nice if `conan search` (or `conan --install` for that matter) had the ability to use a given profile as a query condition, a flag like `--profile=<name>` would make the above much simpler.

**Comment by memsharded** [comment]
Sorry this was not followed up back then.

> As a side note, it would be nice if conan search (or conan --install for that matter) had the ability to use a given profile as a query condition, a flag like --profile=<name> would make the above much simpler.

``conan list --profile`` is being included in next Conan 2.2 release.


Closing this ticket as solved, but please create new ones as necessary, thanks for the feedback!
